feat: Added the enable scorecard docs for IDP (#3620)
harness/developer-hub: docs/internal-developer-portal/features/scorecard.md

---
title: Scorecards

> description: Measure software maturity and best practices

Reviewer: ```suggestion
description: Measure software maturity and enforce best practices
```

:::

### Overview

Scorecards play a pivotal role in ensuring software components are developed and utilized while adhering to organizational and industry standards. They provide a quantifiable measure of software maturity and adherence to best practices, thereby aiding developers in enhancing quality and assisting teams in making informed decisions regarding software adoption.

<docimage width="750vw" path={require('../../internal-developer-portal/features/static/scorecard.png')}/>

<details>
<summary>Purpose & Concept of Scorecards</summary>

- **Measure Software Maturity**: Evaluate the robustness and reliability of software components.
- **Assess Best Practices**: Ensure software adheres to organizational and industry standards.
- **Gamification**: Encourage developers to adhere to standards by providing scores.
- **Confidence Estimation**: Help teams estimate the reliability of software based on its score.

<docimage path={require('../../internal-developer-portal/features/static/concept-scorecard.png')}/>

</details>

:::info

> Few steps like `add proxy` for plugin configuration will be obsolete in the upcoming release and you will be required to add scorecards under the `Scorecard` tab under `Admin` with adequate checks. 

Reviewer: ```suggestion
Few steps like `add proxy` for plugin configuration and layout update will be obsolete in the upcoming releases. They will be handled automatically.
```

::: 

> 1. #### Enable Sccorecards

Reviewer: ```suggestion
1. #### Enable Scorecards
```

>     - **Add Proxy to Harness ci-cd plugin** as mentioned below by navigating to the `Plugin` tab under `Admin` section in the left nav bar. 

Reviewer: CI, CD, IDP, API, JSON, YAML, etc. are all acronyms. They'll always be all caps when used in a sentence. :)

```suggestion
    - **Add Proxy to Harness CI/CD plugin** as mentioned below by navigating to the `Plugin` tab under `Admin` section in the left nav bar. 
```

Reviewer 2: We do not have to mention this info. Proxy config will be autopopulated in the next week's release. Currently, CI-CD plugin configs are non-editable in customer accounts unless we turn ON FF to edit.

Author: We are adding the same info as `Few steps like add proxy for plugin configuration will be obsolete in the upcoming release` and asking the users to enable this FF as well `IDP_ENABLE_EDIT_HARNESS_CI_CD_PLUGIN`.

```
  /harness/scorecard:
    target: https://app.harness.io/
    pathRewrite:
      /api/proxy/harness/scorecard/?: /

>     allowedHeaders:
>     - authorization
>     - Harness-Account

Reviewer: ```suggestion
    allowedHeaders:
      - authorization
      - Harness-Account
```

There should be an indentation here, right? @jenil-harness @sandeepa-kv-jois ?

Author: I have used the example as mentioned in the Layout for CI/CD plugin and I don't think there's any indentation

```

<docimage path={require('../../internal-developer-portal/features/static/admin-ci-cd.png')}/>

<docimage path={require('../../internal-developer-portal/features/static/proxy.png')}/>

### Scorecard Components in IDP

> Scorecard is available in IDP as an independent entity as well as in overview tab as illustrated below. The entity tab contains detailed comprehensive information as shown in the image under [overview](/docs/internal-developer-portal/features/scorecard#overview)

Reviewer: ```suggestion
Scorecard has two main UI components which are developer facing and lives in the Catalog - 1. A small "Card" for the Overview page with the scores and, 2. A "Tab" view with details of the checks and how the score got computed. This is illustrated below. The Tab view contains detailed comprehensive information as shown in the image under [overview](/docs/internal-developer-portal/features/scorecard#overview)
```

<docimage path={require('../../internal-developer-portal/features/static/scorecard-overviewpage.png')}/>

2. #### Adding Card and Tab Content for an Entity

>     - Go to the `Layout` section in IDP `Admin`, and select the **entity kind** for which you want to add scorecards.

Reviewer: ```suggestion
    - Go to the `Layout` section in IDP `Admin`, and select "Service" or any other kind of catalog entity for which you want to add scorecards.
```

3. #### Adding Scorecard Card to the overview tab
    
    - Find the `Overview` tab in the YAML and add the following in its component section.

```
- component: EntityScoreCard
    specs:
    gridProps:
        md: 6

```

4. #### Adding Scorecard Tab content

    - Under the `tabs` section just add the following.

```
- name: Scorecard
  path: /scorecard
  title: Scorecard
  contents:
    - component: EntityScorecardContent
```
<docimage path={require('../../internal-developer-portal/features/static/entity.png')}/>

> 5. #### Removing/Disabling Scorecards

Reviewer: ```suggestion
### Removing/Disabling Scorecards
```

This isn't really a step of the enable flow right? :)

>     - Comment out the lines added under `Layout` tab to remove the scorecard. 

Reviewer: ```suggestion
    - Comment out the Scorecard related lines added under `Layout` tab as mentioned above to remove the scorecard components from the Catalog pages. 
```

>     - Another way to **disable** the scorecard is to convert the same into draft post publishing and it will be disabled as illustrated below. 

Reviewer: ```suggestion
    - Another way to **disable** the scorecards feature is to change the status of all the scorecards you have to "draft". This will ensure that the computation will not run, and will not be shown to the developer.
```

> <docimage width="1750vw" path={require('../../internal-developer-portal/features/static/remove-scorecard.png')}/>

Reviewer: ```suggestion
<docimage width="100vw" path={require('../../internal-developer-portal/features/static/remove-scorecard.png')}/>
```

The `vw` and `vh` units are 0-100. 50vw means, half of available window width. 50vh means half of the loaded page height. https://www.w3schools.com/cssref/css_units.php
